### ISSUE TITLE
Add a RqlCursorEmpty exception to the python driver

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#3974]
Tornado does not play well with `StopIteration` exceptions.  In order to maintain compatibility I would suggest having the following errors:

`RqlCursorEmpty`, inherits from `RqlError`

`DefaultCursorEmpty`, inherits from `RqlCursorEmpty` and `StopIteration`
- this is thrown by `DefaultCursor` during iteration or `next` when the stream is exhausted (but no other errors occurred)

`TornadoCursorEmpty` inherits from `RqlCursorEmpty`
 - this is thrown by `TornadoCursor` futures when yielded and the stream is exhausted (but no other errors occurred)

**Comment by danielmewes** [comment]
Just to make sure I understand the issue:
You mentioned in the other thread
> Unfortunately, if the user doesn't catch this, it interacts poorly with Tornado, which assumes the `StopIteration` means that the coroutine has finished.

Is that a special meaning of `StopIteration` in Tornado? What happens with other exceptions that the user doesn't catch? Would they also terminate the coroutine?

**Comment by Tryneus** [comment]
@danielmewes, yes, you can see in the source [here](http://tornado.readthedocs.org/en/latest/_modules/tornado/gen.html#coroutine), the implementation of `_make_coroutine_wrapper` treats a `StopIteration` the same as a `raise gen.Return(None)`.  Other exceptions will be put into the returned future and will be re-raised when those futures are `yield`ed.

**Comment by danielmewes** [comment]
Got it, thanks!

**Comment by danielmewes** [comment]
Your proposed change makes total sense then. :+1:

**Comment by v3ss0n** [comment]
Leaving a message here to track this.
I am going to test rethinkdb + Tornado  this weekend. 
So for now , user need to catch the ```StopIteration``` right?

**Comment by danielmewes** [comment]
@v3ss0n Yes that's currently the state of things. (you can track threads by clicking the "Subscribe" button on the right by the way)

**Comment by Tryneus** [comment]
This has been approved and merged to `next` in commit 2171061, and cherry-picked into `v2.0.x` in commit c08b40d.  Review was 2742.  Will be in release 2.0.

The proper way to iterate over a TornadoCursor object is now:

```py
while (yield cursor.fetch_next()):
    item = yield cursor.next()
```

or

```py
try:
    while True:
        item = yield cursor.next()
except r.RqlCursorEmpty:
    pass
```

Both `cursor.next()` and `cursor.fetch_next()` take a `wait` optarg to specify how long to wait before timing out (after which a `RqlTimeoutError` exception is raised).  This optarg defaults to `True`, which will wait indefinitely.  `cursor.fetch_next()` only exists on the TornadoCursor implementation.

@v3ss0n, thanks for your interest, let us know if you have any problems or suggestions after trying it out!

**Comment by v3ss0n** [comment]
ah i forgot that. trying now
